src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/WorkspaceDiagnosticsRefresher.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Hosting;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer;
+
+internal sealed class WorkspaceDiagnosticsRefresher : IRazorStartupService, IDisposable
+{
+    private readonly AsyncBatchingWorkQueue _queue;
+    private readonly IProjectSnapshotManager _projectSnapshotManager;
+    private readonly IClientCapabilitiesService _clientCapabilitiesService;
+    private readonly IClientConnection _clientConnection;
+    private bool? _supported;
+    private CancellationTokenSource _disposeTokenSource = new();
+
+    public WorkspaceDiagnosticsRefresher(
+        IProjectSnapshotManager projectSnapshotManager,
+        IClientCapabilitiesService clientCapabilitiesService,
+        IClientConnection clientConnection)
+    {
+        _clientConnection = clientConnection;
+        _projectSnapshotManager = projectSnapshotManager;
+        _clientCapabilitiesService = clientCapabilitiesService;
+        _queue = new(
+            TimeSpan.FromMilliseconds(200),
+            ProcessBatchAsync,
+            _disposeTokenSource.Token);
+        _projectSnapshotManager.Changed += ProjectSnapshotManager_Changed;
+    }
+
+    public void Dispose()
+    {
+        if (_disposeTokenSource.IsCancellationRequested)
+        {
+            return;
+        }
+
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+        _projectSnapshotManager.Changed -= ProjectSnapshotManager_Changed;
```

Comment:
💭 Consider moving this above `_disposeTokenSource.Cancel()` so that no more work can be added before cancelling. If `IProjectSnapshotManager.Updated` fires while the `AsyncBatchingWorkQueue` is being cancelled, it might try to add work, which can result in an `OperationCancelledException`. Not a huge deal, but definitely unneeded.

Reply:
I didn't know adding work threw if the ABWQ cancellation happened. This is good to know , thank you!

Reply:
It's a very tight race, but it's possible to land here: https://github.com/dotnet/razor/blob/9c97001a24394fc4ac0500751190157e109831e5/src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/AsyncBatchingWorkQueue%602.cs#L205